middle_end/closure/closure.ml

```diff
@@ -730,37 +728,34 @@ let is_erasable = function
   | Uclosure _ -> true
   | u -> is_pure u
 
-let rec bind_params_rec loc fpc subst params args body =
-  match (params, args) with
-    ([], []) -> substitute loc fpc subst (Some Int.Map.empty) body
-  | (p1 :: pl, a1 :: al) ->
-      if is_substituable a1 then
-        bind_params_rec loc fpc (V.Map.add (VP.var p1) a1 subst)
-          pl al body
-      else begin
-        let p1' = VP.rename p1 in
-        let u1, u2 =
-          match VP.name p1, a1 with
-          | "*opt*", Uprim(P.Pmakeblock(0, Immutable, kind), [a], dbg) ->
-              a, Uprim(P.Pmakeblock(0, Immutable, kind),
-                       [Uvar (VP.var p1')], dbg)
-          | _ ->
-              a1, Uvar (VP.var p1')
-        in
-        let body' =
-          bind_params_rec loc fpc (V.Map.add (VP.var p1) u2 subst)
-            pl al body in
-        if occurs_var (VP.var p1) body then
-          Ulet(Immutable, Pgenval, p1', u1, body')
-        else if is_erasable a1 then body'
-        else Usequence(a1, body')
-      end
-  | (_, _) -> assert false
-
-let bind_params loc fpc params args body =
+let bind_params ~mutable_vars loc fpc params args body =
```

Comment:
I think that `bind_params` could take the whole environment as well. This is more regular than cherry-picking the part(s) that are required, and easier to upgrade with more environment metadata in the future.